`_flash_3_varlen_hub` backend cannot handle non-contiguous mask

### Describe the bug

Model like Qwen-Image typically has a non-contiguous mask, currently `_flash_3_varlen_hub` backend does not support this, and cause numerical error in forward and backward silently.

### Reproduction

We and produce the bug using the following code snippet

```python
"""
Usage:
    python test_fa3_mask.py
"""

import torch
from diffusers import QwenImageTransformer2DModel

MODEL_CONFIG = dict(
    patch_size=2,
    in_channels=16,
    out_channels=4,
    num_layers=2,
    attention_head_dim=16,
    num_attention_heads=4,
    joint_attention_dim=16,
    guidance_embeds=False,
    axes_dims_rope=(8, 4, 4),
)


def forward(model, hidden_states, encoder_hidden_states, encoder_hidden_states_mask, latent_h):
    inputs = dict(
        hidden_states=hidden_states,
        timestep=torch.full((hidden_states.shape[0],), 0.5, dtype=torch.float32, device=hidden_states.device),
        encoder_hidden_states=encoder_hidden_states,
        encoder_hidden_states_mask=encoder_hidden_states_mask,
        img_shapes=[[(1, latent_h, latent_h)]] * hidden_states.shape[0],
        return_dict=False,
    )
    with torch.no_grad():
        return model(**inputs)[0].float()


def main():
    torch.manual_seed(42)
    device = "cuda"
    dtype = torch.bfloat16

    # One copy with the buggy FA3 varlen backend, one with SDPA as reference.
    model_fa3 = QwenImageTransformer2DModel(**MODEL_CONFIG).to(device, dtype).eval()
    model_ref = QwenImageTransformer2DModel(**MODEL_CONFIG).to(device, dtype).eval()

    model_fa3.set_attention_backend("_flash_3_varlen_hub")
    model_ref.set_attention_backend("native")
    model_ref.load_state_dict(model_fa3.state_dict(), strict=False)

    B, latent_h, T = 2, 4, 8
    H = MODEL_CONFIG["in_channels"]
    D = MODEL_CONFIG["joint_attention_dim"]

    hidden_states = torch.randn(B, latent_h * latent_h, H, device=device, dtype=dtype)
    encoder_hidden_states = torch.randn(B, T, D, device=device, dtype=dtype)

    # 1. Contiguous mask (all tokens valid) — both backends should agree.
    mask_all = torch.ones(B, T, dtype=torch.bool, device=device)
    c_out_fa3 = forward(model_fa3, hidden_states.clone(), encoder_hidden_states.clone(), mask_all, latent_h)
    c_out_ref = forward(model_ref, hidden_states.clone(), encoder_hidden_states.clone(), mask_all, latent_h)
    torch.testing.assert_close(c_out_fa3, c_out_ref, rtol=1e-2, atol=1e-2)

    # 2. Non-contiguous mask (varying text lengths: 2 / 6 out of 8).
    #    QwenImage's joint mask is [text_mask, image_mask], so with masked
    #    text positions the valid token set is NOT a contiguous prefix.
    mask_varlen = torch.zeros(B, T, dtype=torch.bool, device=device)
    mask_varlen[0, :2] = True   # batch 0: only first 2 text tokens
    mask_varlen[1, :6] = True   # batch 1: only first 6 text tokens
    v_out_fa3 = forward(model_fa3, hidden_states.clone(), encoder_hidden_states.clone(), mask_varlen, latent_h)
    v_out_ref = forward(model_ref, hidden_states.clone(), encoder_hidden_states.clone(), mask_varlen, latent_h)
    torch.testing.assert_close(v_out_fa3, v_out_ref, rtol=1e-2, atol=1e-2)


if __name__ == "__main__":
    main()
```

### Logs

```log
AssertionError: Tensor-likes are not close!
```

### System Info

diffusers: main branch

### Who can help?

@sayakpaul 

_No response_

--- a/src/diffusers/models/attention_dispatch.py
+++ b/src/diffusers/models/attention_dispatch.py
@@ -3024,22 +3024,20 @@ def _flash_attention_3_varlen_hub(
 
     if attn_mask is not None:
         attn_mask = _normalize_attn_mask(attn_mask, batch_size, seq_len_kv)
-
-    (_, seqlens_k), (cu_seqlens_q, cu_seqlens_k), (max_seqlen_q, max_seqlen_k) = (
-        _prepare_for_flash_attn_or_sage_varlen(
-            batch_size, seq_len_q, seq_len_kv, attn_mask=attn_mask, device=query.device
+        (_, _), (cu_seqlens_q, cu_seqlens_k), (max_seqlen_q, max_seqlen_k) = (
+            _prepare_for_flash_attn_or_sage_varlen_with_mask(batch_size, seq_len_q, attn_mask, query.device)
         )
-    )
-
-    key_valid, value_valid = [], []
-    for b in range(batch_size):
-        valid_len = seqlens_k[b]
-        key_valid.append(key[b, :valid_len])
-        value_valid.append(value[b, :valid_len])
+        indices_k = attn_mask.flatten().nonzero(as_tuple=False).flatten()
+        key_packed = key.reshape(-1, *key.shape[2:])[indices_k]
+        value_packed = value.reshape(-1, *value.shape[2:])[indices_k]
+    else:
+        (_, _), (cu_seqlens_q, cu_seqlens_k), (max_seqlen_q, max_seqlen_k) = (
+            _prepare_for_flash_attn_or_sage_varlen_without_mask(batch_size, seq_len_q, seq_len_kv, query.device)
+        )
+        key_packed = key.flatten(0, 1)
+        value_packed = value.flatten(0, 1)
 
     query_packed = query.flatten(0, 1)
-    key_packed = torch.cat(key_valid, dim=0)
-    value_packed = torch.cat(value_valid, dim=0)
 
     func = _HUB_KERNELS_REGISTRY[AttentionBackendName._FLASH_3_VARLEN_HUB].kernel_fn
     result = func(

--- a/tests/models/testing_utils/attention.py
+++ b/tests/models/testing_utils/attention.py
@@ -70,8 +70,38 @@ _PARAM_FLASH_3_HUB = pytest.param(
     ],
 )
 
+_PARAM_FLASH_VARLEN_HUB = pytest.param(
+    AttentionBackendName.FLASH_VARLEN_HUB,
+    id="flash_varlen_hub",
+    marks=[
+        pytest.mark.skipif(not _CUDA_AVAILABLE, reason="CUDA is required for flash_varlen_hub backend."),
+        pytest.mark.skipif(
+            not is_kernels_available(),
+            reason="`kernels` package is required for flash_varlen_hub backend. Install with `pip install kernels`.",
+        ),
+    ],
+)
+
+_PARAM_FLASH_3_VARLEN_HUB = pytest.param(
+    AttentionBackendName._FLASH_3_VARLEN_HUB,
+    id="flash_3_varlen_hub",
+    marks=[
+        pytest.mark.skipif(not _CUDA_AVAILABLE, reason="CUDA is required for _flash_3_varlen_hub backend."),
+        pytest.mark.skipif(
+            not is_kernels_available(),
+            reason="`kernels` package is required for _flash_3_varlen_hub backend. Install with `pip install kernels`.",
+        ),
+    ],
+)
+
 # All backends under test.
-_ALL_BACKEND_PARAMS = [_PARAM_NATIVE_CUDNN, _PARAM_FLASH_HUB, _PARAM_FLASH_3_HUB]
+_ALL_BACKEND_PARAMS = [
+    _PARAM_NATIVE_CUDNN,
+    _PARAM_FLASH_HUB,
+    _PARAM_FLASH_3_HUB,
+    _PARAM_FLASH_VARLEN_HUB,
+    _PARAM_FLASH_3_VARLEN_HUB,
+]
 
 # Backends that perform non-deterministic operations and therefore cannot run when
 # torch.use_deterministic_algorithms(True) is active (e.g. after enable_full_determinism()).
@@ -267,6 +297,8 @@ class AttentionBackendTesterMixin:
         Use `pytest -m "not attention"` to skip these tests.
     """
 
+    unsupported_attn_backends: list[str] = []
+
     def setup_method(self):
         gc.collect()
         backend_empty_cache(torch_device)
@@ -280,6 +312,8 @@ class AttentionBackendTesterMixin:
     def test_set_attention_backend_matches_context_manager(self, backend):
         """set_attention_backend() and the attention_backend() context manager must yield identical outputs."""
         _skip_if_backend_requires_nondeterminism(backend)
+        if backend.value in self.unsupported_attn_backends:
+            pytest.skip(f"{backend.value} is not supported for this model.")
 
         init_dict = self.get_init_dict()
         inputs_dict = self.get_dummy_inputs()
@@ -318,6 +352,8 @@ class AttentionBackendTesterMixin:
     def test_output_close_to_native(self, backend, atol=1e-2, rtol=1e-2):
         """All backends should produce model output numerically close to the native SDPA reference."""
         _skip_if_backend_requires_nondeterminism(backend)
+        if backend.value in self.unsupported_attn_backends:
+            pytest.skip(f"{backend.value} is not supported for this model.")
 
         init_dict = self.get_init_dict()
         inputs_dict = self.get_dummy_inputs()
@@ -378,6 +414,8 @@ class AttentionBackendTesterMixin:
         as opposed to `model.compile`).
         """
         _skip_if_backend_requires_nondeterminism(backend)
+        if backend.value in self.unsupported_attn_backends:
+            pytest.skip(f"{backend.value} is not supported for this model.")
         if getattr(self.model_class, "_repeated_blocks", None) is None:
             pytest.skip("Skipping tests as regional compilation is not supported.")
 

--- a/tests/models/testing_utils/utils.py
+++ b/tests/models/testing_utils/utils.py
@@ -8,6 +8,7 @@ _BF16_REQUIRED_BACKENDS = {
     AttentionBackendName.FLASH_HUB,
     AttentionBackendName.FLASH_VARLEN_HUB,
     AttentionBackendName._FLASH_3_HUB,
+    AttentionBackendName._FLASH_3_VARLEN_HUB,
 }
 
 

--- a/tests/models/transformers/test_models_transformer_qwenimage.py
+++ b/tests/models/transformers/test_models_transformer_qwenimage.py
@@ -22,6 +22,7 @@ from diffusers.utils.torch_utils import randn_tensor
 
 from ...testing_utils import enable_full_determinism, torch_device
 from ..testing_utils import (
+    AttentionBackendTesterMixin,
     AttentionTesterMixin,
     BaseModelTesterConfig,
     BitsAndBytesTesterMixin,
@@ -248,6 +249,21 @@ class TestQwenImageTransformerTraining(QwenImageTransformerTesterConfig, Trainin
 
 class TestQwenImageTransformerAttention(QwenImageTransformerTesterConfig, AttentionTesterMixin):
     """Attention processor tests for QwenImage Transformer."""
+
+
+class TestQwenImageTransformerAttentionBackend(QwenImageTransformerTesterConfig, AttentionBackendTesterMixin):
+    """Attention backend tests for QwenImage Transformer."""
+
+    unsupported_attn_backends = ["flash_hub", "_flash_3_hub"]
+
+    def get_dummy_inputs(self, batch_size: int = 2):
+        inputs = super().get_dummy_inputs(batch_size=batch_size)
+        mask = inputs["encoder_hidden_states_mask"]
+        mask[...] = 0
+        mask[0, :2] = 1
+        mask[1, :6] = 1
+        inputs["encoder_hidden_states_mask"] = mask.bool()
+        return inputs
 
 
 class TestQwenImageTransformerContextParallel(QwenImageTransformerTesterConfig, ContextParallelTesterMixin):

